### PR TITLE
✨Add Validation Before Submit

### LINKED ISSUE #13: ♻️Validate all form fields before submit
#### Describe your issue

To prevent corrupted data, we should validate all form fields before the user can submit/ continue.

#### Possible solution

Add a loop to validate all form fields, before the user can submit them.

#### Issue checklist

Please check the boxes in this list after submitting your Issue:

- [x] I've checked if this issue already exists
- [x] I've included all the information that i think is relevant
- [x] I've added logs and/or screenshots (if applicable)
- [x] I've mentioned PRs and issues that relate to this one

### PR DESCRIPTION
**Pull request by MaFa961**
**Changes:**

1. Add a validation before the user can submit the Usertask results
1. Add visual validation error

- To support empty form fields, the variable `isValid` is by default true.

**Issues:**

Closes #13 

PR: #15 

---

See the [examples](#example) for inspiration.

## How can others test the changes?

Fill out some form fields with wrong input, like a date with `30.02.2020`.

## PR-Checklist

Please check the boxes in this list after submitting your PR:

- [x] You can merge this PR **right now** (if not, please prefix the title with "WIP: ")
- [x] I've tested **all** changes included in this PR.
- [x] I've also reviewed this PR myself before submitting (e.g. for scrambled letters, typos, etc.).
- [x] I've rebased the `develop` branch with my branch before finishing this PR.
- [x] I've **summarized all changes** in a list above.
- [x] I've mentioned all **PRs, which relate to this one**.
- [x] I've prefixed my Pull Request title is according to [gitmoji guide](https://gitmoji.carloscuesta.me/).

### CODE REVIEW

File: src/components/usertask-component/dynamic-usertask-component.tsx [hunk 1]
```diff
@@ -160,6 +164,16 @@ export class DynamicUserTaskComponent {
     this.canceled.emit();
   }
 
+  private _isInputValid(): boolean {
+    for (const formField of this._formFields) {
+      if (!formField.isValid) {
```
Comment (alexanderkasten) [inline src/components/usertask-component/dynamic-usertask-component.tsx]: please extract business rule

Comment (Vyperus) [inline src/components/usertask-component/dynamic-usertask-component.tsx]:
```suggestion
      if (!formField.isValid) {
      
```

File: src/components/usertask-component/dynamic-usertask-component.tsx [hunk 2]
```diff
@@ -127,13 +127,17 @@ export class DynamicUserTaskComponent {
...
     event.preventDefault();
 
-    this.submitted.emit({
-      correlationId: this.usertask.correlationId,
-      processInstanceId: this.usertask.processInstanceId,
-      userTaskId: this.usertask.id,
-      userTaskInstanceId: this.usertask.flowNodeInstanceId,
-      results: this._getFormResults(),
-    });
+    const isValid: boolean = this._isInputValid();
```
Comment (robpal94) [inline src/components/usertask-component/dynamic-usertask-component.tsx]:
```suggestion
    const inputIsValid: boolean = this._isInputValid();
```

File: src/components/usertask-component/dynamic-usertask-component.tsx [hunk 3]
```diff
@@ -127,13 +127,17 @@ export class DynamicUserTaskComponent {
...
+
+    if (isValid) {
```
Comment (robpal94) [inline src/components/usertask-component/dynamic-usertask-component.tsx]:
```suggestion
    if (inputIsValid) {
```

File: src/components/usertask-component/form-fields/date-form-field.tsx [hunk 4]
```diff
@@ -35,13 +36,23 @@ export class DateFormField implements IFormField {
...
+    this.isValid = this._inputValidator.isValidDate(event.target.value);
+
+    if (this.isValid) {
+      const element: HTMLElement = document.getElementById(this.formField.id);
+      element.style.borderColor = '';
+    } else {
+      const element: HTMLElement = document.getElementById(this.formField.id);
+      element.style.borderColor = 'red';
+    }
+
```
Comment (robpal94) [inline src/components/usertask-component/form-fields/date-form-field.tsx]: I would simplify this a bit:

```TypeScript
const element: HTMLElement = document.getElementById(this.formField.id);
element.style.borderColor = this.isValid ? 'red' : '';
```

File: src/components/usertask-component/form-fields/long-form-field.tsx [hunk 5]
```diff
@@ -2,6 +2,7 @@ import {DataModels} from '@process-engine/consumer_api_contracts';
 import {Component, State} from '@stencil/core';
 
 import {IFormField} from './iform_field';
+import { IKeyDownOnInputEvent } from './ikey_down_on_input_event';
```
Comment (Vyperus) [inline src/components/usertask-component/form-fields/long-form-field.tsx]:
```suggestion
import {IKeyDownOnInputEvent} from './ikey_down_on_input_event';
```

File: src/components/usertask-component/form-fields/number-form-field.tsx [hunk 6]
```diff
@@ -35,10 +36,16 @@ export class NumberFormField implements IFormField {
               <label>{this.formField.label}</label>
               <input type='text' class='form-control' id={this.formField.id} name={this.formField.label}
                 placeholder='0.0' value={this.value} pattern='^(-?\d+(,|\.)\d+)|(\d+)$'
-                onKeyDown={(event: any): void => this._handleKeyDown(event)} onInput={(event: any): void => this._handleInput(event)}></input>
+                onKeyDown={(event: any): void => this._handleKeyDown(event)} onInput={(event: any): void => this._handleInput(event)}
+                onChange={(event: any): void => this._handleChange(event)}></input>
```
Comment (Vyperus) [inline src/components/usertask-component/form-fields/number-form-field.tsx]:
```suggestion
                onChange={(event: any): void => this._handleChange(event)}>
              </input>
```

File: src/components/usertask-component/form-fields/long-form-field.tsx [hunk 7]
```diff
@@ -33,11 +35,17 @@ export class LongFormField implements IFormField {
     return <div class='form-group'>
               <label htmlFor={this.formField.id}>{this.formField.label}</label>
               <input type='text' class='form-control' id={this.formField.id} name={this.formField.label} value={this.value}
-                placeholder='0' pattern='^\d+$'
-                onKeyDown={(event: any): void => this._handleKeyDown(event)} onInput={(event: any): void => this._handleInput(event)}></input>
+                placeholder='0' pattern='^\d*$'
+                onKeyDown={(event: any): void => this._handleKeyDown(event)} onInput={(event: any): void => this._handleInput(event)}
+                onChange={(event: any): void => this._handleChange(event)}> </input>
```
Comment (Vyperus) [inline src/components/usertask-component/form-fields/long-form-field.tsx]:
```suggestion
                onChange={(event: any): void => this._handleChange(event)}>
              </input>
```

File: src/components/usertask-component/form-fields/date-form-field.tsx [hunk 8]
```diff
@@ -35,13 +36,28 @@ export class DateFormField implements IFormField {
             <label htmlFor={this.formField.id}>{this.formField.label}</label>
             <input type='text' data-provide='datepicker' class='form-control' maxlength='10' placeholder='--.--.----'
             pattern='^(0?[1-9]|[12][0-9]|3[01])([ \.])(0?[1-9]|1[012])\2([0-9][0-9][0-9][0-9])(([ -])([0-1]?[0-9]|2[0-3]):[0-5]?[0-9]:[0-5]?[0-9])?$'
-            id={this.formField.id} value={this.value} onFocus={(event: any): void => this._handleChange(event)}
+            id={this.formField.id} value={this.value} onChange={(event: any): void => this._handleChange(event)}
             onKeyDown={(event: any): void => this._handleKeyDown(event)}></input>
```
Comment (Vyperus) [inline src/components/usertask-component/form-fields/date-form-field.tsx]:
```suggestion
            onKeyDown={(event: any): void => this._handleKeyDown(event)}>
          </input>
```

File: src/components/usertask-component/form-fields/long-form-field.tsx [hunk 9]
```diff
@@ -48,6 +57,16 @@ export class LongFormField implements IFormField {
...
   }
 
+  private _setStyle(event: IKeyDownOnInputEvent): void {
+    const isEmptyInput: boolean = event.target.value.length === 0;
+
+    const element: HTMLElement = document.getElementById(this.formField.id);
+    element.style.borderColor = (this.isValid || isEmptyInput) ? '' : 'red';
+    if (isEmptyInput) {
+      this.isValid = true;
+    }
```
Comment (robpal94) [inline src/components/usertask-component/form-fields/long-form-field.tsx]: I also would simplify this:

```TypeScript
this.isValid = isEmtpyInput
```

File: src/components/usertask-component/dynamic-usertask-component.tsx [hunk 10]
```diff
@@ -160,6 +164,18 @@ export class DynamicUserTaskComponent {
     this.canceled.emit();
   }
 
+  private _isInputValid(): boolean {
+    for (const formField of this._formFields) {
+      const hasFormFieldValidInput: boolean = formField.isValid;
```
Comment (robpal94) [inline src/components/usertask-component/dynamic-usertask-component.tsx]:
```suggestion
      const formFieldInputIsInvalid: boolean = !formField.isValid;
```

File: src/components/usertask-component/dynamic-usertask-component.tsx [hunk 11]
```diff
@@ -160,6 +164,18 @@ export class DynamicUserTaskComponent {
...
+      if (!hasFormFieldValidInput) {
```
Comment (robpal94) [inline src/components/usertask-component/dynamic-usertask-component.tsx]:
```suggestion
      if (formFieldInputIsInvalid) {
```

File: src/components/usertask-component/form-fields/date-form-field.tsx [hunk 12]
```diff
@@ -34,14 +35,30 @@ export class DateFormField implements IFormField {
     return <div class= 'form-group'>
```
Comment (robpal94) [inline src/components/usertask-component/form-fields/date-form-field.tsx]:
```suggestion
    const validationRegexpr: string = 
      '^(0?[1-9]|[12][0-9]|3[01])([ \.])(0?[1-9]|1[012])\2([0-9][0-9][0-9][0-9])(([ -])([0-1]?[0-9]|2[0-3]):[0-5]?[0-9]:[0-5]?[0-9])?$';
    return <div class= 'form-group'>
```

File: src/components/usertask-component/form-fields/date-form-field.tsx [hunk 13]
```diff
@@ -34,14 +35,30 @@ export class DateFormField implements IFormField {
...
             <label htmlFor={this.formField.id}>{this.formField.label}</label>
             <input type='text' data-provide='datepicker' class='form-control' maxlength='10' placeholder='--.--.----'
-            pattern='^(0?[1-9]|[12][0-9]|3[01])([ \.])(0?[1-9]|1[012])\2([0-9][0-9][0-9][0-9])(([ -])([0-1]?[0-9]|2[0-3]):[0-5]?[0-9]:[0-5]?[0-9])?$'
-            id={this.formField.id} value={this.value} onFocus={(event: any): void => this._handleChange(event)}
-            onKeyDown={(event: any): void => this._handleKeyDown(event)}></input>
+             pattern='^(0?[1-9]|[12][0-9]|3[01])([ \.])(0?[1-9]|1[012])\2([0-9][0-9][0-9][0-9])(([ -])([0-1]?[0-9]|2[0-3]):[0-5]?[0-9]:[0-5]?[0-9])?$'
```
Comment (robpal94) [inline src/components/usertask-component/form-fields/date-form-field.tsx]:
```suggestion
             pattern={validationRegexpr}
```

File: src/components/usertask-component/form-fields/date-form-field.tsx [hunk 14]
```diff
@@ -34,14 +35,30 @@ export class DateFormField implements IFormField {
...
+            </input>
           </div>;
   }
 
   private _handleChange(event: IKeyDownOnInputEvent): void {
     this.value = event.target.value;
+    this.isValid = this._inputValidator.isValidDate(event.target.value);
+
+    this._setStyle(event);
+
```
Comment (robpal94) [inline src/components/usertask-component/form-fields/date-form-field.tsx]:
```suggestion
```

File: src/components/usertask-component/form-fields/long-form-field.tsx [hunk 15]
```diff
@@ -33,11 +35,18 @@ export class LongFormField implements IFormField {
     return <div class='form-group'>
               <label htmlFor={this.formField.id}>{this.formField.label}</label>
               <input type='text' class='form-control' id={this.formField.id} name={this.formField.label} value={this.value}
-                placeholder='0' pattern='^\d+$'
-                onKeyDown={(event: any): void => this._handleKeyDown(event)} onInput={(event: any): void => this._handleInput(event)}></input>
+                placeholder='0' pattern='^\d*$'
```
Comment (robpal94) [inline src/components/usertask-component/form-fields/long-form-field.tsx]: See above. I would always extract the pattern into a dedicated variable. This way, you can easier find and read the associated regex.

File: src/components/usertask-component/form-fields/date-form-field.tsx [hunk 16]
```diff
@@ -34,14 +35,30 @@ export class DateFormField implements IFormField {
     return <div class= 'form-group'>
             <label htmlFor={this.formField.id}>{this.formField.label}</label>
             <input type='text' data-provide='datepicker' class='form-control' maxlength='10' placeholder='--.--.----'
-            pattern='^(0?[1-9]|[12][0-9]|3[01])([ \.])(0?[1-9]|1[012])\2([0-9][0-9][0-9][0-9])(([ -])([0-1]?[0-9]|2[0-3]):[0-5]?[0-9]:[0-5]?[0-9])?$'
-            id={this.formField.id} value={this.value} onFocus={(event: any): void => this._handleChange(event)}
-            onKeyDown={(event: any): void => this._handleKeyDown(event)}></input>
+             pattern='^(0?[1-9]|[12][0-9]|3[01])([ \.])(0?[1-9]|1[012])\2([0-9][0-9][0-9][0-9])(([ -])([0-1]?[0-9]|2[0-3]):[0-5]?[0-9]:[0-5]?[0-9])?$'
+             id={this.formField.id} value={this.value} onChange={(event: any): void => this._handleChange(event)}
```
Comment (alexanderkasten) [inline src/components/usertask-component/form-fields/date-form-field.tsx]:
```suggestion
             id={this.formField.id} value={this.value} onChange={(event: KeyboardEvent): void => this._handleChange(event)}
```

File: src/components/usertask-component/form-fields/date-form-field.tsx [hunk 17]
```diff
@@ -34,14 +35,30 @@ export class DateFormField implements IFormField {
...
+             onKeyDown={(event: any): void => this._handleKeyDown(event)}>
```
Comment (alexanderkasten) [inline src/components/usertask-component/form-fields/date-form-field.tsx]:
```suggestion
             onKeyDown={(event: KeyboardEvent): void => this._handleKeyDown(event)}>
```

File: src/components/usertask-component/form-fields/long-form-field.tsx [hunk 18]
```diff
@@ -33,11 +35,18 @@ export class LongFormField implements IFormField {
...
+                onKeyDown={(event: any): void => this._handleKeyDown(event)} onInput={(event: any): void => this._handleInput(event)}
```
Comment (alexanderkasten) [inline src/components/usertask-component/form-fields/long-form-field.tsx]:
```suggestion
                onKeyDown={(event: KeyboardEvent): void => this._handleKeyDown(event)} onInput={(event: KeyboardEvent): void => this._handleInput(event)}
```

File: src/components/usertask-component/form-fields/long-form-field.tsx [hunk 19]
```diff
@@ -33,11 +35,18 @@ export class LongFormField implements IFormField {
...
+                onChange={(event: any): void => this._handleChange(event)}>
```
Comment (alexanderkasten) [inline src/components/usertask-component/form-fields/long-form-field.tsx]:
```suggestion
                onChange={(event: IKeyDownOnInputEvent): void => this._handleChange(event)}>
```

File: src/components/usertask-component/form-fields/number-form-field.tsx [hunk 20]
```diff
@@ -35,10 +36,17 @@ export class NumberFormField implements IFormField {
               <label>{this.formField.label}</label>
               <input type='text' class='form-control' id={this.formField.id} name={this.formField.label}
                 placeholder='0.0' value={this.value} pattern='^(-?\d+(,|\.)\d+)|(\d+)$'
-                onKeyDown={(event: any): void => this._handleKeyDown(event)} onInput={(event: any): void => this._handleInput(event)}></input>
+                onKeyDown={(event: any): void => this._handleKeyDown(event)} onInput={(event: any): void => this._handleInput(event)}
+                onChange={(event: any): void => this._handleChange(event)}>
```
Comment (alexanderkasten) [inline src/components/usertask-component/form-fields/number-form-field.tsx]:
```suggestion
                onChange={(event: IKeyDownOnInputEvent): void => this._handleChange(event)}>
```

File: src/components/usertask-component/form-fields/number-form-field.tsx [hunk 21]
```diff
@@ -35,10 +36,17 @@ export class NumberFormField implements IFormField {
               <label>{this.formField.label}</label>
               <input type='text' class='form-control' id={this.formField.id} name={this.formField.label}
                 placeholder='0.0' value={this.value} pattern='^(-?\d+(,|\.)\d+)|(\d+)$'
-                onKeyDown={(event: any): void => this._handleKeyDown(event)} onInput={(event: any): void => this._handleInput(event)}></input>
+                onKeyDown={(event: any): void => this._handleKeyDown(event)} onInput={(event: any): void => this._handleInput(event)}
```
Comment (alexanderkasten) [inline src/components/usertask-component/form-fields/number-form-field.tsx]:
```suggestion
                onKeyDown={(event: KeyboardEvent): void => this._handleKeyDown(event)} onInput={(event: KeyboardEvent): void => this._handleInput(event)}
```

**Pull request merged**